Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -520,9 +523,31 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 
     PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(timebase.get(), timerSource.get(), nextPurgeTime, 0);
 
+    m_nextScheduledPurge = nextPurgeTime;
+
     return samplesPurged;
 }
 
+void VideoMediaSampleRenderer::maybeReschedulePurge()
+{
+    assertIsCurrent(dispatcher().get());
+
+    if (!decodedSamplesCount())
+        return;
+    if (!m_nextScheduledPurge) {
+        purgeDecodedSampleQueue(ShouldReschedule::Yes);
+        return;
+    }
+    RetainPtr nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()));
+    ASSERT(nextSample);
+
+    CMTime duration = PAL::CMSampleBufferGetOutputDuration(nextSample.get());
+    CMTime presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp(nextSample.get());
+    CMTime presentationEndTime = PAL::CMTimeAdd(presentationTime, duration);
```

Comment:
I would be tempted to add a `shouldPergeSample()` that would do this check.
Then, we would write it:
```
if (!shouldHeadSample())
    return;
purgeDecodedSampleQueue(ShouldReschedule::Yes);
```